### PR TITLE
feat: add `FileUpload` component

### PR DESCRIPTION
**Pull request by snitin315**
## Description

- API - #2010 
- Tests - #2040

### CODE REVIEW

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: image

The delete icon is going outside when file name is large

Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: Is validation of files needs to be handled on consumer end?

I dragged and dropped a PDF file and it was able to capture and upload it

image

Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: I think accept prop is not working for drag & drop. Are you able to select a pdf file from on clicking upload?

Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: It should have been truncated. Can you share the file with me on slack?

Screenshot 2024-02-23 at 2 32 33 PM

**Comment by anuraghazra** [comment]
There's a quite a lot of layout shift which is happening after selecting a file, this is happening because after the "Upload GST" label the margin is increasing, ideally margin should stay same. 

https://github.com/razorpay/blade/assets/35374649/eb71436a-8f61-43b1-b83a-6e8338c0f864


On design the margin stays consistent: 

image

**Comment by anuraghazra** [comment]
File name and the Tick icon doesn't look aligned vertically. 

image

On design: 

image

**Comment by anuraghazra** [comment]
There's a gap in the bottom after the progressbar: 

image

Gap not present in design: 

image

**Comment by anuraghazra** [comment]
In the first storybook example, can you wire up the controls? I cannot play around with different variants like isDisabled.

**Comment by anuraghazra** [comment]
Check once with RK on how it looks in mobile: 

image

Looks a bit too cramped, we may need to add vertical padding in that Upload CTA

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx
Comment (saurabhdaware) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: no it works fine when I click and select. But not working on drag and drop

**Comment by anuraghazra** [comment]
Tried on my android, truncation fails and text overflows for some reason: 

image

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx
Comment (saurabhdaware) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: sent file on slack. Maybe because the filename didn't have spacing in my case

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+  const validateFiles = (inputFiles: BladeFileList, allFiles: BladeFileList): boolean => {
+    if (uploadType === 'single' && inputFiles.length > 1) {
+      setErrorMessage('You can upload only one file.');
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (maxCount && allFiles.length > maxCount) {
+      setErrorMessage(`You can't upload more than ${maxCount} files.`);
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:104]: These messages are hardcoded? 

What if product wants to change these?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+          </FormLabel>
+        ) : null}
+
+        <BaseBox display="flex" flexDirection="column" marginBottom="spacing.5">
+          <SelectorLabel
+            componentName={MetaConstants.FileUploadLabel}
+            inputProps={{}}
+            style={{
+              cursor: isDisabled ? 'not-allowed' : 'pointer',
+              ...(!isMultiple && selectedFiles.length === 1 ? screenReaderStyles : {}),
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:189]: `!isMultiple && selectedFiles.length === 1`

Factor out and name this conditional with a variable.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+            />
+          )}
+        </BaseBox>
+
+        {selectedFiles.map((file) => (
+          <MemoizedFileUploadItem
+            key={file.id}
+            file={file}
+            onRemove={() => {
+              const newFiles = selectedFiles.filter(({ id }) => id !== file.id);
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:300]: Move this before line 296?

File: packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          fill={getIn(theme.colors, 'interactive.icon.negative.disabled')}
+        />
+      </Svg>
+    );
+  }
+
+  const FILE_EXTENSION_REGEX = /(?:\.([^.]+))?$/i;
+  const [fileExtension, fileType] = FILE_EXTENSION_REGEX.exec(fileName) ?? [];
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx:32]: Do we need regex? 

Can we not split the file in "." and get the last element? 🤔

File: packages/blade/src/components/Form/Selector/SelectorInput.web.tsx [hunk 5]
```diff
@@ -30,7 +30,8 @@ const getHoverStyles = ({
 
   return {
     borderColor: borderColor ? getIn(theme, borderColor) : undefined,
-    backgroundColor: getIn(theme, backgroundColor),
+    backgroundColor:
+      backgroundColor === 'transparent' ? 'transparent' : getIn(theme, backgroundColor),
```
Comment (anuraghazra) [inline packages/blade/src/components/Form/Selector/SelectorInput.web.tsx:34]: We don't need this. 

We have `color.transparent` token in our theme file now.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    invalid: Boolean(showError),
+    disabled: Boolean(isDisabled),
+    describedBy: helpTextId,
+  });
+
+  // In control mode attach a unique id to each file if not provided
+  useMemo(() => {
+    for (const file of selectedFiles) {
+      if (!file.id) {
+        file.id = `${new Date().getTime().toString()}${Math.floor(Math.random() * 1000000)}`;
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:80]: We are mutating the files without setState?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText =
+    !isMultiple && selectedFiles.length === 1 ? false : Boolean(helpText) || Boolean(errorMessage);
+
+  const showError = validationState === 'error' || internalValidationState === 'error';
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
+    accessibilityLabel ?? `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileinput');
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:67]:
```suggestion
  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileuploadinput');
```

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+import { assignWithoutSideEffects } from '~utils/assignWithoutSideEffects';
+import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
+import { getStyledProps } from '~components/Box/styledProps';
+import { makeSize, useTheme } from '~utils';
+import { Text } from '~components/Typography';
+import type { BladeElementRef } from '~utils/types';
+import { getHintType } from '~components/Input/BaseInput/BaseInput';
+import { makeAccessible } from '~utils/makeAccessible';
+
+const MemoizedFileUploadItem = memo(FileUploadItem);
```
Comment (anuraghazra) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:22]: Why do we need to memoize it here? 

Can we not just export a memoized FileInput if this is needed?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 6 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:80]: If we use setState it will go in an infinite re-render loop.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:1]: @saurabhdaware Fixed

**Comment by snitin315** [comment]
> File name and the Tick icon doesn't look aligned vertically.

Fixed

> In the first storybook example, can you wire up the controls? I cannot play around with different variants like isDisabled.


Done

> Check once with RK on how it looks in mobile:

Updated 

>Tried on my android, truncation fails and text overflows for some reason:

Can you try again? I think this is the same as https://github.com/razorpay/blade/pull/2016#discussion_r1500426797

>There's a gap in the bottom after the progressbar:

Fixed.

@anuraghazra

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 8 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx]: yes, done.

File: packages/blade/src/components/Form/Selector/SelectorInput.web.tsx (hunk 5 above)
Comment (snitin315) [reply packages/blade/src/components/Form/Selector/SelectorInput.web.tsx]: fixed

File: packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    const easing = getIn(theme.motion, fileUploadMotionTokens.easing);
+    const duration = castWebType(
+      makeMotionTime(getIn(theme.motion, fileUploadMotionTokens.duration)),
+    );
+
+    return {
+      display: 'flex',
+      justifyContent: 'space-between',
+      borderStyle: 'solid',
+      minHeight: '56px',
```
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx]: fixed

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 2 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx]: done

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 1 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:111]: These are the default validations messages, giving users control to customize this might complex the logic. We can improve if any use-case comes up. Additional validations & messages can be set via onChange & errorText props.

**Comment by anuraghazra** [comment]
> File name and the Tick icon doesn't look aligned vertically.

Weird, cannot find any stories which shows the tick mark anymore, Unable to confirm this. 

> There's a gap in the bottom after the progressbar:

The file upload progress [story](https://61c19ee8d3d282003ac1d81c-ekmmhfqlla.chromatic.com/?path=/story/components-fileupload--auto-file-upload-with-progress) doesn't show the progressbar anymore upon selecting the file, Unable to confirm this.

**Comment by anuraghazra** [comment]
this [one](https://github.com/razorpay/blade/pull/2016#issuecomment-1960989807) still exists.

**Comment by anuraghazra** [comment]
labelPosition: left looks broken: 

image

**Comment by snitin315** [comment]
> The file upload progress [story](https://61c19ee8d3d282003ac1d81c-ekmmhfqlla.chromatic.com/?path=/story/components-fileupload--auto-file-upload-with-progress) doesn't show the progressbar anymore upon selecting the file, Unable to confirm this.

My bad! I've fixed the stories now.

>labelPosition: left looks broken:

Currently, It is the same as other input labels.

Screenshot 2024-02-26 at 3 28 15 PM

// @anuraghazra

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+  const [internalValidationState, setInternalValidationState] = useState('none');
+  const [isActive, setIsActive] = useState(false);
+
+  const isMultiple = uploadType === 'multiple';
+  const isOneFileSelectedWithSingleUpload = !isMultiple && selectedFiles.length === 1;
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText = isOneFileSelectedWithSingleUpload
+    ? false
+    : Boolean(helpText) || Boolean(errorMessage);
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:62]: We dont render helpText in mutli upload?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+                  setSelectedFiles(newFiles);
+                  onCancel?.({ file: selectedFiles[0] });
+                }}
+                onPreview={onPreview}
+              />
+            </BaseBox>
+          )}
+        </BaseBox>
+      </BaseBox>
+      {/* the magic number 136 is basically max-width of label i.e 120 and then right margin i.e 16 which is the spacing between label and input field */}
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:314]: Move it constant maybe?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 1 above)
Comment (saurabhdaware) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:111]: When users pass errorText, will this error message be overriden? or both error messages exist at different places?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+    const hasValidationErrors = validateFiles(droppedFiles, allFiles);
+
+    if (!hasValidationErrors) {
+      handleFilesChange(droppedFiles);
+      onDrop?.({ name, fileList: allFiles });
+    }
+  };
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    const inputFiles = Array.from(event.target.files ?? []);
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:153]: is it not already an array?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+                onDragOver={handleDragOver}
+                onDragLeave={handleDragLeave}
+                onDrop={handleDrop}
+                onClick={() => setIsActive(true)}
+              >
+                <Box
+                  display="flex"
+                  justifyContent="center"
+                  alignItems="center"
+                  flexDirection={{ xs: 'column', s: 'row' }}
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:229]:
```suggestion
                  flexDirection={{ base: 'column', s: 'row' }}
```

Is this expected btw? so we want it to be flex on smaller phones than normal `s` size phones?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+                      ...accessibilityProps,
+                    }}
+                    ref={ref}
+                  />
+
+                  <Box
+                    display="flex"
+                    justifyContent="center"
+                    alignItems="center"
+                    flexDirection={{ xs: 'column', s: 'row' }}
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:265]:
```suggestion
                    flexDirection={{ base: 'column', s: 'row' }}
```

Better to use `base` wherever you're using responsive props. It acts as a default value when value is not defined

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+              onPreview={onPreview}
+            />
+          </BaseBox>
+        ))}
+    </BaseBox>
+  );
+};
+
+const FileUpload = assignWithoutSideEffects(forwardRef(_FileUpload), {
+  displayName: MetaConstants.FileUpload,
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:362]: It shouldn't be MetaConstants no? we usually have TitleCased names here for displayName and componentId

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+    }
+
+    if (maxCount && allFiles.length > maxCount) {
+      setErrorMessage(`You can't upload more than ${maxCount} files.`);
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (maxSize && inputFiles.some((file) => file.size > maxSize)) {
+      setErrorMessage('File size exceeded.');
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:117]: do we not need to tell file size exceeded for which file?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+
+  const validateFiles = (inputFiles: BladeFileList, allFiles: BladeFileList): boolean => {
+    if (accept && inputFiles.some((file) => !isFileAccepted(file, accept))) {
+      setErrorMessage(`You provided an unsupported file type. Supported file types are: ${accept}`);
+      setInternalValidationState('error');
+      return true;
+    }
+
+    if (uploadType === 'single' && inputFiles.length > 1) {
+      setErrorMessage('You can upload only one file.');
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:105]: Check these error messages with design also once if they are not coming from design already

File: packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    case 'docx':
+      iconColors.background = globalColors.chromatic.azure[50];
+      iconColors.border = globalColors.chromatic.azure[600];
+      iconColors.cover = globalColors.chromatic.azure[700];
+      iconColors.fold = globalColors.chromatic.azure[800];
+      break;
+    case 'xls':
+    case 'xlsx':
+    case 'csv':
+      iconColors.background = globalColors.chromatic.emerald[50];
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUploadItemIcon.tsx:57]: are we using global colors on design also?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+  const isOneFileSelectedWithSingleUpload = !isMultiple && selectedFiles.length === 1;
+  const inputLabelPosition = platform === 'onMobile' ? 'top' : labelPosition;
+  const isLabelLeftPositioned = inputLabelPosition === 'left';
+  const willRenderHintText = isOneFileSelectedWithSingleUpload
+    ? false
+    : Boolean(helpText) || Boolean(errorMessage);
+
+  const showError = validationState === 'error' || internalValidationState === 'error';
+  const showHelpText = !showError && helpText;
+  const accessibilityText =
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:66]: Shouldn't the label get priority when accesisibilityLabel is not present?

Like right now when announcing component, it is only announcing help text `".png, .jpg" should be uploaded` but there is no context to users on what type of image. That context will be on label no?

image

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+    accessibilityLabel ?? `,${showError ? errorText : ''} ${showHelpText ? helpText : ''}`;
+  const { inputId, labelId, helpTextId, errorTextId } = useFormId('fileuploadinput');
+
+  const accessibilityProps = makeAccessible({
+    required: Boolean(isRequired),
+    invalid: Boolean(showError),
+    disabled: Boolean(isDisabled),
```
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:73]: They can be `undefined` too.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 19 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:66]: I'll check this.

File: packages/blade/src/components/Box/BaseBox/types/propsTypes.ts [hunk 21]
```diff
@@ -127,7 +127,7 @@ type CommonBoxVisualProps = MakeObjectResponsive<
   {
     borderRadius: keyof Border['radius'];
     borderWidth: keyof Border['width'];
-    borderColor: BorderColorString<'surface'>;
+    borderColor: BorderColorString<'surface'> | BorderColorString<'interactive'>;
```
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/blade/src/components/Box/BaseBox/types/propsTypes.ts:130]: We support drag & drop props on Box too so I think we should support it on both.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 1 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:111]: `errorText` will override these.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 6 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:81]: No, hooks can't be called inside a callback.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 15 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:362]: Correct, fixed.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 14 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:265]: updated.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 13 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:229]: yes,

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 12 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:153]: No, [`FileList`](https://developer.mozilla.org/en-US/docs/Web/API/FileList) is not an array.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 16 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:115]: Thinking more on this, we don't need this info. There can be multiple files with long names too. We can add it later if requested by users.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 19 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:64]: Fixed.

**Comment by snitin315** [comment]
@anuraghazra Fixed the stories, you can check again.

> this https://github.com/razorpay/blade/pull/2016#issuecomment-1960989807 still exists.

Fixed

> labelPosition: left looks broken:
> 

It's the current behavior of FormLabel for all our inputs. The `*` goes into the next line.
https://github.com/razorpay/blade/blob/fe64b50e12d66a2eae2580f4b2b109e121067c59/packages/blade/src/components/Form/FormLabel.tsx#L113
Screenshot 2024-02-28 at 8 26 07 AM

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 11 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:310]: It's used in Baseinpu & OTPInput as well, let's refactor this separately.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 10 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx]: We do, fixed.

File: packages/blade/src/components/Box/BaseBox/types/propsTypes.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (snitin315) [reply packages/blade/src/components/Box/BaseBox/types/propsTypes.ts]: For now, I refactored code to avoid using `borderColor` prop. but let's allow `interactive` token for Box in separate PR because we support Drag & Drop.

[the remark this answers is not in the document]
Comment (saurabhdaware) [reply packages/blade/src/components/Box/BaseBox/types/propsTypes.ts]: makes sense

File: packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx [hunk 22]
```diff
@@ -24,6 +24,7 @@ const StyledFileUploadItemWrapper = styled(BaseBox)<StyledFileUploadItemWrapperP
       transitionProperty: 'background-color',
       transitionTimingFunction: easing,
       transitionDuration: duration,
+      borderColor: getIn(theme.colors, 'interactive.border.neutral.faded'),
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx:27]:
```suggestion
      borderColor: theme.colors.interactive.border.neutral.faded,
```

This also works in this case right?

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+              }}
+              onPreview={onPreview}
+            />
+          </BaseBox>
+        ))}
+    </BaseBox>
+  );
+};
+
+const FileUpload = assignWithoutSideEffects(forwardRef(_FileUpload), {
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:357]: Do add jsdoc here with basic usage

**Comment by saurabhdaware** [comment]
The examples in stories and examples on stackblitz seem like same examples. Maybe you can keep basic examples in stories and move complex examples in stackblitz because they'll be able to see full code on stackblitz

**Comment by snitin315** [comment]
> The examples in stories and examples on stackblitz seem like same examples. Maybe you can keep basic examples in stories and move complex examples in stackblitz because they'll be able to see full code on stackblitz

Stackblitz takes significant boot-up time hence I've kept both. Users can play around with stories and look at examples if they want to see the implementation.

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx (hunk 23 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:470]: Done.

File: packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx (hunk 22 above)
Comment (snitin315) [reply packages/blade/src/components/FileUpload/StyledFileUploadItemWrapper.tsx]: Yes, updated.

**Review by anuraghazra — APPROVED** [review]

File: packages/blade/src/components/FileUpload/FileUpload.web.tsx [hunk 24]
```diff
@@ -354,6 +354,119 @@ const _FileUpload: React.ForwardRefRenderFunction<BladeElementRef, FileUploadPro
   );
 };
 
+/**
+ * ### FileUpload Component
+ *
```
Comment (saurabhdaware) [inline packages/blade/src/components/FileUpload/FileUpload.web.tsx:359]: This is a little too large for jsdoc 🙈 Its hard to parse through this much code in small dialog box of jsdoc. Just keep `<FileUpload />` component with its props maybe? ignore the state management and surrounding box buttons etc. Doesn't have to be a working example

Comment (snitin315) [reply packages/blade/src/components/FileUpload/FileUpload.web.tsx:359]: Updated but in https://github.com/razorpay/blade/pull/2040

**Review by saurabhdaware — APPROVED** [review]
